packages/shared-auth/mode/localhost.ts

```diff
@@ -0,0 +1,60 @@
+import { Elysia, status } from "elysia";
+import isLocalhostIp from "is-localhost-ip";
+
+declare global {
+  var __HEBO_WARNED_LOCALHOST: boolean | undefined;
+}
+
+const warnLocalhostOnce = () => {
+  if (globalThis.__HEBO_WARNED_LOCALHOST) return;
+  console.warn(
+    '⚠️ [auth] Localhost mode: userId="dummy"; non-local requests will be 403',
+  );
+  globalThis.__HEBO_WARNED_LOCALHOST = true;
+};
+
+const getRequestHostname = () =>
+  new Elysia({ name: "get-request-hostname" })
+    .derive(({ request }) => {
+      const hostHeader = request.headers.get("host");
+      const raw = (hostHeader ?? "").trim();
+      if (!raw) return { hostname: "" } as const;
+      try {
+        const hostname = new URL(`http://${raw}`).hostname;
+        return { hostname } as const;
+      } catch {
+        return { hostname: raw } as const;
+      }
+    })
+    .as("scoped");
+
+const getClientIp = () =>
+  new Elysia({ name: "get-client-ip" })
+    .derive(({ server, request }) => {
```

Comment:
IP variable is only used by one middleware, why the complexity of a derive?

Have you tried to use request.ip instead?